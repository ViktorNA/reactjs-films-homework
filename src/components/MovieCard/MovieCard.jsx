import propTypes from 'prop-types';
import React, { Component } from 'react';
import Button from '../Buttons';
import styles from './MovieCard.scss';

const cn = require('classnames');
const genreIds = require('../../assets/json/genres.json').genres;

const baseUrl = 'https://image.tmdb.org/t/p/w500/';

export default class MovieCard extends Component {
  renderGenreList = (genres = []) => genres.map(id => genreIds.find(el => el.id === id).name).slice(0, 3).join(', ');

  changeMode = () => {
    const { updater, id, activeId } = this.props;
    updater(id === activeId ? -1 : id);
  }

  checkVisability = (className) => {
    const { id, activeId } = this.props;
    return id === activeId ? className : styles.hidden;
  }

  getBackStyle = () => {
    const { id, activeId } = this.props;
    return cn(styles.back, { [styles.description]: id === activeId });
  }

  render() {
    const {
      movie: {
        title,
        overview,
        poster_path: imageUrl,
        vote_average: rating,
        genre_ids: genres,
        id,
        activeId,
      },
    } = this.props;
    return (
      <div className={styles.card}>
        <img className={styles.imgWrapper} src={baseUrl + imageUrl} alt="movie logo" />
        <div className={styles.trailerBox}>
          <div className={styles.circle}>
            <div className={styles.triangle} />
          </div>
          <div className={styles.label}>Watch Now</div>
          <Button type="viewInfo" text="View Info" action={this.changeMode} />
        </div>
        <div className={this.getBackStyle()}>
          <div className={this.checkVisability(styles.closeWrapper)}>
            <button type="button" className={styles.close} onClick={this.changeMode} />
          </div>
          <div className={styles.movieName}>{title}</div>
          <div className={styles.ratingBox}>{rating}</div>
          <div className={styles.genresBox}>{this.renderGenreList(genres)}</div>
          <div className={this.checkVisability(styles.descriptionWrapper)}>{overview}</div>
          <Button type={cn({ watchNow: id === activeId }, { hidden: id !== activeId })} text="Watch Now" action={() => {}} />
        </div>
      </div>
    );
  }
}

MovieCard.propTypes = {
  movie: propTypes.shape({
    title: propTypes.string.isRequired,
    overview: propTypes.string.isRequired,
    imageUrl: propTypes.string,
    vote_average: propTypes.number.isRequired,
    genres: propTypes.array,
  }),
  id: propTypes.number.isRequired,
  activeId: propTypes.number.isRequired,
  updater: propTypes.func.isRequired,
};

MovieCard.defaultProps = {
  movie: {
    title: 'error',
    overview: 'error',
    vote_average: 0,
    genres: ['uncnown genre'],
  },
};
