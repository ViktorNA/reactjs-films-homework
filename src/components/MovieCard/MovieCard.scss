.card {
  overflow: hidden;
  display: block;
  position: relative;
  width: 250px;
  height: 300px;
  background-color: #b3b3b3;
  box-shadow: 0 5px 20px 0 rgba(0, 0, 0, 0.3);
  border-radius: 5px;
}

.imgWrapper {
  position: absolute;
  width: inherit;
  height: inherit;
  object-fit: cover;
}

.back {
  position: absolute;
  top: 100;
  left: 0;
  right: 0;
  bottom: 0%;
  z-index: 2;
  padding: 20px;
  width: inherit;
  overflow: hidden;
  box-sizing: border-box;
  transition: all 0.5s;
  height: 80px;
  color: black;
  background-color: white;
}

.movieName {
  display: inline-block;
  font-size: 14px;
  font-weight: bold;
  text-align: left;
  width: 150px;
  background-color: rgba(255, 255, 255, 0);
  color: inherit;
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
}

.ratingBox {
  display: inline-block;
  font-size: 12px;
  text-align: center;
  margin-left: 24px;
  width: 28px;
  border: 1px solid #47c3f0;
  border-radius: 3px;
  color: #47c3f0;
  position: absolute;
}

.genresBox {
  font-size: 11px;
  color: #47c3f0;
}

.trailerBox {
  position: relative;
  height: 100%;
  background-color: rgba(0, 0, 0, 0.64);
  z-index: 2;
  display: grid;
  grid-template: 1fr 1fr 1fr 80px/1fr auto 1fr;
  justify-items: center;
  padding-top: 35px;
  opacity: 0;
  transition: all 0.5s;
}

.trailerBox:hover {
  opacity: 1;
}

.triangle {
  z-index: 2;
  width: 0;
  height: 0;
  border-top: 16px solid transparent;
  border-left: 20px solid white;
  border-bottom: 16px solid transparent;
  position: absolute;
  bottom: 15px;
  left: 23px;
}

.circle {
  display: grid;
  grid-column: 2/3;
  position: relative;
  width: 60px;
  height: 60px;
  background: #3bb4e0;
  border-radius: 50%;
}

.label {
  display: grid;
  grid-column: 2/3;
  font-size: 14px;
  color: white;
  margin-top: 10px;
}

.description {
  color: white;
  background-color: rgba(0, 0, 0, 0.64);
  height: 100%;
}

.descriptionWrapper {
  overflow: hidden;
  font-size: 13px;
  margin-top: 20px;
  line-height: 1.5em;
  height: 7.5em;
}

.hidden {
  display: none;
}

.close {
  background-color: rgba(255, 255, 255, 0);
  color: rgba(255, 255, 255, 0);
  height: 15px;
  width: 15px;
  border: none;
  cursor: pointer;
}

.close::before,
.close::after {
  position: absolute;
  left: 6px;
  top: 0;
  content: ' ';
  height: 18px;
  width: 1px;
  background-color: white;
}

.close::before {
  transform: rotate(45deg);
}

.close::after {
  transform: rotate(-45deg);
}

.closeWrapper {
  display: block;
  position: relative;
  margin-bottom: 25px;
}
