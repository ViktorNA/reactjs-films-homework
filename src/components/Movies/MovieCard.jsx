import propTypes from 'prop-types';
import React, { Component } from 'react';
import styles from './MovieCard.scss';

export default class MovieCard extends Component {
  state = {
    infoType: 'short',
  }

  getBackStyle() {
    const { infoType } = this.state;
    return infoType === 'short' ? styles.backShort : styles.backLong;
  }

  getCloseStyle() {
    const { infoType } = this.state;
    return infoType === 'long' ? styles.closeWrapper : styles.hiddenDesc;
  }

  getWatchStyle() {
    const { infoType } = this.state;
    return infoType === 'long' ? styles.watchNowButton : styles.hiddenDesc;
  }

  getHeadStyle() {
    const { infoType } = this.state;
    return infoType === 'video' ? styles.shortHead : styles.head;
  }

  getGenresList = genres => [...genres].slice(0, 3).join(', ');

  setInfoType = (event) => {
    this.setState({ infoType: event.target.name });
  }

  showInfo(description) {
    const { infoType } = this.state;
    switch (infoType) {
      case 'video': return (
        <div className={styles.watchWrapper}>
          <div className={styles.circle}><div className={styles.triangle} /></div>
          <div className={styles.label}>Watch Now</div>
          <button name="long" type="button" onClick={this.setInfoType} className={styles.viewInfoButton}>
            View Info
          </button>
        </div>
      );
      case 'short': return (
        <div />
      );
      default: break;
    }
    return (
      <div className={styles.visibleDesc}>
        {description}
      </div>
    );
  }

  render() {
    const {
      movie: {
        name,
        description,
        imageUrl,
        rating,
        genres,
      },
    } = this.props;
    return (
      <div className={styles.card}>
        <img className={styles.imgWrapper} src={imageUrl} alt="Movie log" />
        <div className={this.getBackStyle()}>
          <div className={this.getCloseStyle()}>
            <button name="short" type="button" className={styles.close} onClick={this.setInfoType}>x</button>
          </div>
          <div className={this.getHeadStyle()}>
            <button name="long" className={styles.movieName} type="button" onClick={this.setInfoType}>
              {name}
            </button>
            <div className={styles.ratingBox}>
              {rating}
            </div>
            <div className={styles.genresBox}>
              {this.getGenresList(genres)}
            </div>
          </div>
          {this.showInfo(description)}
          <button name="video" type="button" className={this.getWatchStyle()} onClick={this.setInfoType}>
            Watch Now
          </button>
        </div>
      </div>
    );
  }
}

MovieCard.propTypes = {
  movie: propTypes.shape({
    name: propTypes.string.isRequired,
    description: propTypes.string.isRequired,
    imageUrl: propTypes.string,
    rating: propTypes.number.isRequired,
    genres: propTypes.array,
  }),
};

MovieCard.defaultProps = {
  movie: {
    name: 'error',
    description: 'error',
    rating: 0,
    genres: ['uncnown genre'],
  },
};
