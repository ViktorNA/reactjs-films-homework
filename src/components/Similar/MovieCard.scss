.card {
  overflow: hidden;
  display: block;
  position: relative;
  width: 250px;
  height: 300px;
  background-color: red;
  border-radius: 5px;
}

.description {
  font-size: 13px;
  line-height: 1.6em;
  position: relative;
  height: 100px;
  margin-top: 19px;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
  overflow: hidden;
  color: white;
}

.visibleDesc {
  composes: description;
  display: block;
}

.hiddenDesc {
  composes: description;
  display: none;
}

.imgWrapper {
  position: absolute;
  width: inherit;
  height: inherit;
  background-size: cover;
  background-position: center;
}

.back {
  position: absolute;
  top: 100;
  left: 0;
  right: 0;
  bottom: 0%;
  z-index: 999;
  padding: 20px;
  width: inherit;
  overflow: hidden;
  box-sizing: border-box;
  transition: all 1s;
}

.backShort {
  composes: back;
  height: 80px;
  color: black;
  background-color: white;
}

.backLong {
  composes: back;
  height: inherit;
  background-color: rgba(0, 0, 0, 0.64);
  color: white;
}

.close {
  background-color: rgba(255, 255, 255, 0);
  color: rgba(255, 255, 255, 0);
  height: 15px;
  width: 15px;
  border: none;
}

.close::before,
.close::after {
  position: absolute;
  left: 6px;
  top: 0;
  content: ' ';
  height: 18px;
  width: 1px;
  background-color: white;
}

.close::before {
  transform: rotate(45deg);
}

.close::after {
  transform: rotate(-45deg);
}

.movieName {
  font-size: 14px;
  text-align: left;
  width: 150px;
  border: 0;
  outline: 0;
  background-color: rgba(255, 255, 255, 0);
  color: inherit;
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
  padding: 0;
}

.closeWrapper {
  position: relative;
  margin-bottom: 25px;
}

.ratingBox {
  display: inline-block;
  font-size: 12px;
  text-align: center;
  margin-left: 24px;
  width: 28px;
  border: 1px solid #47c3f0;
  border-radius: 3px;
  color: #47c3f0;
}

.genresBox {
  font-size: 11px;
  color: #47c3f0;
}

.watchNowButton {
  font-size: 14px;
  color: white;
  border: 0;
  border-radius: 3px;
  background-color: #3bb4e0;
  width: 100px;
  height: 30px;
  margin-top: 23px;
}

.triangle {
  z-index: 999;
  width: 0;
  height: 0;
  border-top: 16px solid transparent;
  border-left: 20px solid white;
  border-bottom: 16px solid transparent;
  position: absolute;
  bottom: 15px;
  left: 23px;
}

.circle {
  display: grid;
  grid-column: 2/3;
  position: relative;
  width: 60px;
  height: 60px;
  background: #3bb4e0;
  border-radius: 50%;
}

.watchWrapper {
  display: grid;
  grid-template-columns: 1fr auto 1fr;
  justify-items: center;
  margin-top: 35px;
}

.label {
  display: grid;
  grid-column: 2/3;
  font-size: 14px;
  margin-top: 10px;
}

.head {
  display: block;
}

.shortHead {
  position: absolute;
  bottom: 0;
  left: 0;
  padding: 20px;
  box-sizing: border-box;
  height: 80px;
  width: inherit;
  background-color: white;
  color: black;
}

.viewInfoButton {
  grid-column: 2/3;
  margin-top: 27px;
  font-size: 14px;
  color: white;
  border-radius: 3px;
  border: 1px solid white;
  width: 100px;
  height: 28px;
  background-color: rgba(255, 255, 255, 0);
  align-content: center;
}
