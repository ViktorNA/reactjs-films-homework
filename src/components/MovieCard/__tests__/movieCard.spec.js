import TestRenderer from 'react-test-renderer';
import React from 'react';
import MovieCard from '../MovieCard';

describe('Snapshot test', () => {
  it('works correctly', () => {
    const testRender = TestRenderer.create(<MovieCard id={12} activeId={12} updater={() => { }} />);
    expect(testRender.toJSON()).toMatchSnapshot();
  });
});

describe('genres to string', () => {
  it('works correctly', () => {
    const testRender = TestRenderer.create(<MovieCard id={12} activeId={12} updater={() => { }} />);
    const testInstance = testRender.getInstance();
    const res = testInstance.renderGenreList([12]);
    expect(res).toEqual('Adventure');
  });
});

describe('changeMode', () => {
  it('reset mode', () => {
    const updater = jest.fn();
    const testRender = TestRenderer.create(<MovieCard id={12} activeId={12} updater={updater} />);
    const testInstance = testRender.getInstance();
    testInstance.changeMode();
    expect(updater).toHaveBeenCalledWith(-1);
  });
  it('set mode', () => {
    const updater = jest.fn();
    const testRender = TestRenderer.create(<MovieCard id={12} activeId={1} updater={updater} />);
    const testInstance = testRender.getInstance();
    testInstance.changeMode();
    expect(updater).toHaveBeenCalledWith(12);
  });
});

describe('checkVisability', () => {
  it('reset mode', () => {
    const testRender = TestRenderer.create(<MovieCard id={12} activeId={12} />);
    const testInstance = testRender.getInstance();
    const res = testInstance.checkVisability('className');
    expect(res).toBe('className');
  });
  it('set mode', () => {
    const testRender = TestRenderer.create(<MovieCard id={12} activeId={1} />);
    const testInstance = testRender.getInstance();
    const res = testInstance.checkVisability('className');
    expect(res).toBe('hidden');
  });
});

describe('getBackStyle', () => {
  it('desc', () => {
    const testRender = TestRenderer.create(<MovieCard id={12} activeId={12} />);
    const testInstance = testRender.getInstance();
    const res = testInstance.getBackStyle();
    expect(res).toBe('back description');
  });
  it('short', () => {
    const testRender = TestRenderer.create(<MovieCard id={12} activeId={1} />);
    const testInstance = testRender.getInstance();
    const res = testInstance.getBackStyle();
    expect(res).toBe('back');
  });
});
