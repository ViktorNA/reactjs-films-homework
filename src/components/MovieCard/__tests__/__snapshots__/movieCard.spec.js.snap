// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot test works correctly 1`] = `
<div
  className="card"
>
  <img
    alt="movie logo"
    className="imgWrapper"
    src="https://image.tmdb.org/t/p/w500/undefined"
  />
  <div
    className="trailerBox"
  >
    <div
      className="circle"
    >
      <div
        className="triangle"
      />
    </div>
    <div
      className="label"
    >
      Watch Now
    </div>
    <button
      className="viewInfo common"
      onClick={[Function]}
      style={Object {}}
      type="button"
    >
      View Info
    </button>
  </div>
  <div
    className="back description"
  >
    <div
      className="closeWrapper"
    >
      <button
        className="close"
        onClick={[Function]}
        type="button"
      />
    </div>
    <div
      className="movieName"
    >
      error
    </div>
    <div
      className="ratingBox"
    >
      0
    </div>
    <div
      className="genresBox"
    >
      
    </div>
    <div
      className="descriptionWrapper"
    >
      error
    </div>
    <button
      className="watchNow common"
      onClick={[Function]}
      style={Object {}}
      type="button"
    >
      Watch Now
    </button>
  </div>
</div>
`;
